kotlinx-coroutines-core/common/test/flow/operators/TimeoutTest.kt

```diff
@@ -99,16 +99,19 @@ class TimeoutTest : TestBase() {
     @Test
     fun testUpstreamErrorTimeoutException() = testUpstreamError(TimeoutCancellationException(0, Job()))
 
+    @Test
+    fun testUpstreamErrorCancellationException() = testUpstreamError(CancellationException(""))
+
     private inline fun <reified T: Throwable> testUpstreamError(cause: T) = runTest {
         try {
             // Workaround for JS legacy bug
             flow {
                 emit(1)
                 throw cause
-            }.timeout(1.milliseconds).collect()
+            }.timeout(1000.milliseconds).collect()
             expectUnreached()
         } catch (e: Throwable) {
-            assertTrue { e is T }
+            assertIs<T>(e)
             finish(1)
         }
     }
```

Comment:
You raise a good point about a race not being needed. I think though that we should also check what happens in case of a race between the `flow` block throwing an exception and the `timeout` directive. WDYT?
```suggestion
    }

    @Test
    fun testUpstreamExceptionsTakingPriority() = withVirtualTime {
        val flow = flow<Unit> {
            expect(2)
            withContext(NonCancellable) {
                delay(2.milliseconds)
            }
            assertFalse(currentCoroutineContext().isActive) // cancelled already
            expect(3)
            throw TestException()
        }.timeout(1.milliseconds)
        expect(1)
        assertFailsWith<TestException> {
            flow.collect {
                expectUnreached()
            }
        }
        finish(4)
    }
```

Reply:
Makes sense, let's check it as well